### PR TITLE
Add capability of creating a subdomain

### PR DESCRIPTION
**Pull request by Kebree**
##### SUMMARY
Add feature for win_domain to create a new domain in a forest instead of creating a new forest.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_domain

##### ADDITIONAL INFORMATION
Until now, win_domain is only capable of creating a new forest. With this feature, you can add a new domain into an existing forest.
The module is also switched from Ansible Legacy framework to SharpUtil Ansible.Basic with better arguments parsing

### CODE REVIEW

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 1]
```diff
@@ -24,25 +24,55 @@ Function Ensure-Prereqs {
...
-$safe_mode_admin_password = Get-AnsibleParam -obj $params -name "safe_mode_password" -failifempty $true
-$database_path = Get-AnsibleParam -obj $params -name "database_path" -type "path"
-$sysvol_path = Get-AnsibleParam -obj $params -name "sysvol_path" -type "path"
-$create_dns_delegation = Get-AnsibleParam -obj $params -name "create_dns_delegation" -type "bool"
-$domain_mode = Get-AnsibleParam -obj $params -name "domain_mode" -type "str"
-$forest_mode = Get-AnsibleParam -obj $params -name "forest_mode" -type "str"
+$spec = @{
+    options = @{
+        dns_domain_name = @{ type = "str"; required = $true }
+        domain_netbios_name = @{ type = "str"; required = $false }
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain.ps1]: `required = $false` is the default and doesn't need to be set

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 2]
```diff
@@ -24,25 +24,55 @@ Function Ensure-Prereqs {
...
+        safe_mode_password = @{ type = "str"; no_log = $true; required = $true }
+        database_path = @{ type = "path" }
+        sysvol_path = @{ type = "path" }
+        create_dns_delegation = @{ type = "bool" }
+        domain_mode = @{ type = "str" }
+        forest_mode = @{ type = "str" }
+        parent_domain_name = @{ type = "str" }
+        parent_domain_user = @{ type = "str"; default="Administrator" }
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain.ps1]: I would leave off the default and make people explicitly put it there.

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 3]
```diff
@@ -24,25 +24,55 @@ Function Ensure-Prereqs {
...
+$spec = @{
+    options = @{
+        dns_domain_name = @{ type = "str"; required = $true }
+        domain_netbios_name = @{ type = "str"; required = $false }
+        safe_mode_password = @{ type = "str"; no_log = $true; required = $true }
+        database_path = @{ type = "path" }
+        sysvol_path = @{ type = "path" }
+        create_dns_delegation = @{ type = "bool" }
+        domain_mode = @{ type = "str" }
+        forest_mode = @{ type = "str" }
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain.ps1]: We should be implementing our choices validation here. I know we do a further check below but that's mostly to verify we don't set something like `WinThreshold` on a Server 2012 R2 host.

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 4]
```diff
@@ -24,25 +24,55 @@ Function Ensure-Prereqs {
...
-$forest_mode = Get-AnsibleParam -obj $params -name "forest_mode" -type "str"
+$spec = @{
+    options = @{
+        dns_domain_name = @{ type = "str"; required = $true }
+        domain_netbios_name = @{ type = "str"; required = $false }
+        safe_mode_password = @{ type = "str"; no_log = $true; required = $true }
+        database_path = @{ type = "path" }
+        sysvol_path = @{ type = "path" }
+        create_dns_delegation = @{ type = "bool" }
+        domain_mode = @{ type = "str" }
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain.ps1]: Same thing about `forest_mode`.

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 5]
```diff
@@ -56,99 +86,188 @@ $installed = Ensure-Prereqs
...
-        SafeModeAdministratorPassword=$sm_cred;
-        Confirm=$false;
-        SkipPreChecks=$true;
-        InstallDns=$true;
-        NoRebootOnCompletion=$true;
-        WhatIf=$check_mode;
+    # Check that we got a valid forest_mode
+    $valid_forest_modes = [Enum]::GetNames((Get-Command -Name Install-ADDSForest).Parameters.ForestMode.ParameterType)
+    if (($null -ne $forest_mode) -and -not ($forest_mode -in $valid_forest_modes)) {
+        $module.FailJson("The parameter 'forest_mode' does not accept '$forest_mode', please use one of: $valid_forest_modes")
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain.ps1]: It might be better to say, this Windows Version does not support this forest mode to better reflect the check that is happening.

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 6]
```diff
@@ -56,99 +86,188 @@ $installed = Ensure-Prereqs
...
+
+    $domain = $null
+    try {
+        $domain_context =  New-Object -TypeName System.DirectoryServices.ActiveDirectory.DirectoryContext -ArgumentList Domain, $domain_fqdn
+        $domain = [System.DirectoryServices.ActiveDirectory.Domain]::GetDomain($domain_context)
+    } catch [System.DirectoryServices.ActiveDirectory.ActiveDirectoryObjectNotFoundException] {
+    } catch [System.DirectoryServices.ActiveDirectory.ActiveDirectoryOperationException] { }
+
+    if (-not $domain) {
+        $result.changed = $true
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_domain.ps1]: There's a lot of shared code here, it would be great if you could have the code below part of a function that each if statement calls.

**Review by jborean93 — CHANGES_REQUESTED** [review]
Looks good, would like to see some of the duplicated code in a shared function and the choices added back in.

**Comment by Kebree** [comment]
ready_for_review

**Comment by Kebree** [comment]
ready_for_review

**Comment by Kebree** [comment]
@ansibot ready_for_review

**Comment by Kebree** [comment]
bot_status

File: lib/ansible/modules/windows/win_domain.py [hunk 7]
```diff
@@ -71,6 +72,26 @@
     type: str
     choices: [ Win2003, Win2008, Win2008R2, Win2012, Win2012R2, WinThreshold ]
     version_added: '2.8'
+  parent_domain_name:
+    description:
+    - Specifies the parent domain while creating a new domain in a forest.
+    - Has to be the parent domain's fqdn
```
Comment (ShachafGoldstein) [inline lib/ansible/modules/windows/win_domain.py]: Seems silly but all the other lines have a `.` at the end

File: lib/ansible/modules/windows/win_domain.py [hunk 8]
```diff
@@ -71,6 +72,26 @@
...
+    - It can't be used with forest mode as we don't create a new forest.
```
Comment (ShachafGoldstein) [inline lib/ansible/modules/windows/win_domain.py]: Please look [here](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-within-module-documentation) to make the reference for the parameter of `forest_mode`

File: lib/ansible/modules/windows/win_domain.py [hunk 9]
```diff
@@ -71,6 +72,26 @@
...
+    type: str
+    version_added: '2.9'
+  parent_domain_user:
+    description:
+    - Credential of the parent domain.
+    - The user has to be entreprise administrator to create a new subdomain.
+    - Required if parent_domain_name is defined.
```
Comment (ShachafGoldstein) [inline lib/ansible/modules/windows/win_domain.py]: Please look [here ](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-within-module-documentation)to make the reference for the parameter

File: lib/ansible/modules/windows/win_domain.py [hunk 10]
```diff
@@ -71,6 +72,26 @@
...
+    type: str
+    version_added: '2.9'
+  parent_domain_password:
+    description:
+    - Password of the parent domain's entreprise administrator.
+    - Required if parent_domain_name is defined.
```
Comment (ShachafGoldstein) [inline lib/ansible/modules/windows/win_domain.py]: Please look [here ](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-within-module-documentation)to make the reference for the parameter

File: lib/ansible/modules/windows/win_domain.py [hunk 11]
```diff
@@ -106,4 +127,18 @@
...
     sysvol_path: C:\Windows\SYSVOL
   register: domain_install
+
+- name: Create new Windows domain in an existing forest with specific parameters
+  win_domain:
+    create_dns_delegation: yes
+    database_path: C:\Windows\NTDS
+    dns_domain_name: child
+    domain_mode: Win2012R2
+    domain_netbios_name: ANSIBLE-CHILD
```
Comment (ShachafGoldstein) [inline lib/ansible/modules/windows/win_domain.py]: I think it would be better to not have an example of NetBIOS name different than the DNS name. I know the above has the same issue but regarding new additions, I don't think we want to encourage this usage.

File: lib/ansible/modules/windows/win_domain.ps1 [hunk 12]
```diff
@@ -2,7 +2,7 @@
 
 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
-#Requires -Module Ansible.ModuleUtils.Legacy
+#AnsibleRequires -CSharpUtil Ansible.Basic
```
Comment (ShachafGoldstein) [inline lib/ansible/modules/windows/win_domain.ps1]: This is not only a feature PR since you actually changed the module to use the new C# utils.
This should be a different PR, one for the utils and one for the feature.
I know it sounds like a lot but it will make the reviews faster and be the better way to add all of the changes into the module.

**Comment by jborean93** [comment]
Hi @Kebree are you still interested in trying to get this PR into Ansible? If you are it just needs a rebase and we can start the review process again. I'm trying to clear out as much of the backlog of the Windows PRs so let me know your thoughts.

**Comment by Kebree** [comment]
Looks like I failed... 
I have my code on another computer, I'll do it again monday.

**Comment by jborean93** [comment]
Awesome, feel free to ping me when you have one ready and I’ll do the review as soon as I can.
